src/main/java/com/snowflake/kafka/connector/internal/ingestsdk/StreamingClientManager.java

```diff
@@ -0,0 +1,146 @@
+ */
+package com.snowflake.kafka.connector.internal.ingestsdk;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.LoggerHandler;
+import com.snowflake.kafka.connector.internal.SnowflakeErrors;
+import com.snowflake.kafka.connector.internal.streaming.StreamingUtils;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Provides access to the streaming ingest clients. This should be the only place to manage clients.
+ */
+public class StreamingClientManager {
+  private LoggerHandler LOGGER;
+
+  private Map<Integer, KcStreamingIngestClient> taskToClientMap;
+  private int maxTasks;
+  private final int minTasks = 0;
+
+  // TESTING ONLY - inject the client map
+  @VisibleForTesting
+  public StreamingClientManager(Map<Integer, KcStreamingIngestClient> taskToClientMap) {
+    this();
+    this.taskToClientMap = taskToClientMap;
+  }
+
+  /** Creates a new client manager */
+  protected StreamingClientManager() {
+    LOGGER = new LoggerHandler(this.getClass().getName());
+    this.taskToClientMap = new HashMap<>();
+    this.maxTasks = 0;
+  }
+
+  /**
+   * Creates as many clients as needed with the connector config and kc instance id. This assumes
+   * that all taskIds are consecutive ranging from 0 to maxTasks.
+   *
+   * @param connectorConfig the config for the clients, cannot be null
+   * @param kcInstanceId the kafka connector id requesting the clients, cannot be null
+   * @param maxTasks the max number of tasks assigned to this connector, must be greater than 0
+   * @param numTasksPerClient the max number of tasks to be assigned to each client, must be greater
+   *     than 0
+   */
+  public void createAllStreamingClients(
+      Map<String, String> connectorConfig,
+      String kcInstanceId,
+      int maxTasks,
+      int numTasksPerClient) {
+    assert connectorConfig != null && kcInstanceId != null && maxTasks > 0 && numTasksPerClient > 0;
+
+    this.maxTasks = maxTasks;
+
+    int clientCount = (int) Math.ceil((double) maxTasks / (double) numTasksPerClient);
+    LOGGER.info(
+        "Creating {} clients for {} tasks with max {} tasks per client",
+        clientCount,
+        maxTasks,
+        numTasksPerClient);
+
+    Properties clientProperties = new Properties();
+    clientProperties.putAll(
+        StreamingUtils.convertConfigForStreamingClient(new HashMap<>(connectorConfig)));
+
+    // put a new client for every tasksToCurrClient taskIds
+    int taskId = 0;
+    int clientId = 0;
+    int tasksToCurrClient = 0;
+    KcStreamingIngestClient createdClient =
+        new KcStreamingIngestClient(
+            clientProperties,
+            KcStreamingIngestClient.buildStreamingIngestClientName(kcInstanceId, clientId));
+    while (taskId < maxTasks) {
+      if (tasksToCurrClient == numTasksPerClient) {
+        tasksToCurrClient = 1;
+        clientId++;
+        createdClient =
```

Comment:
may be add a logger here everytime you create another client?

Reply:
added debug level log, since we have a info level log outside the loop. And refactored code to be a little cleaner